backend/api/src/api/v1/gql/chat.rs

```diff
@@ -0,0 +1,80 @@
+use super::error::{GqlError, Result};
+use super::models::message::Message;
+use crate::global::GlobalState;
+use async_graphql::{
+    futures_util::{Stream, StreamExt},
+    Context, Object, Subscription,
+};
+use async_stream::stream;
+use chrono::Utc;
+use fred::prelude::PubsubInterface;
+use std::sync::Arc;
+
+#[derive(Default, Clone)]
+pub struct ChatMutation;
+
+#[derive(Default, Clone)]
+pub struct ChatSubscription;
+
+#[Object]
+impl ChatMutation {
+    // Send message in chat. You need to be logged in for that.
+    async fn send_message<'ctx>(
+        &self,
+        ctx: &Context<'_>,
+        #[graphql(desc = "ID of chat room where the message will be send.")] chat_id: i64,
+        #[graphql(desc = "Message content that will be published.")] content: String,
+    ) -> Result<bool> {
+        let global = ctx
+            .data::<Arc<GlobalState>>()
+            .expect("failed to get global state")
+            .clone();
+
+        match global.redis_pool.publish::<String, String, String>(chat_id.to_string(), serde_json::json!({ "chat_id": chat_id, "content": content, "message_type": "message" }).to_string()).await {
+            Ok(_) => {},
+            Err(_) => {
+                return Err(GqlError::InternalServerError.with_message("Failed to publish message"));
+            }
+        };
+
+        let _ = sqlx::query!(
+            "INSERT INTO chat_messages (chat_room_id, author_id, message, created_at) VALUES ($1, $2, $3, $4)",
+            chat_id,
+            1,
+            content,
+            Utc::now()
+        )
+        .execute(&global.db)
+        .await;
+
+        Ok(true)
+    }
+}
+
+#[Subscription]
+impl ChatSubscription {
+    // Listen to new messages in chat.
+    pub async fn new_message(
+        &self,
+        ctx: &Context<'_>,
+        #[graphql(desc = "Chat to subscribe to.")] id: String,
+    ) -> impl Stream<Item = Message> {
+        let global = ctx
+            .data::<Arc<GlobalState>>()
+            .expect("failed to get global state")
+            .clone();
+
+        let redis_conn = global.redis_sub_pool.next();
+        let mut message_stream = redis_conn.on_message();
+        let _ = redis_conn.subscribe(id).await;
```

Comment:
I believe fred has a subscription helper, with this impl each subscription will require its own redis connection when in reality you can just resubscribe to more topics. Rather dont use a redis pool for subs but instead a single connection.

Reply:
Fred has some helpers which make this super easy